syntax = "proto3";

import "google/protobuf/descriptor.proto";
import "google/protobuf/timestamp.proto";
import "labels.proto";

option cc_enable_arenas = true;

// Custom Message type
message Vector2 {
  float x = 1;
  float y = 2;
}

message Vector3 {
  float x = 1;
  float y = 2;
  float z = 3;
}

message BoundingBox {
  Vector3 position = 1;
  Vector3 size = 2;
  float yaw = 3;
}

message PolygonBox {
  repeated Vector2 points = 1;
  float min_z = 2;
  float max_z = 3;
}

message TrackingInfo {
  float probability = 4;
  bool tracking_reliable = 5;
  Vector3 velocity = 6;
  repeated Vector3 history = 7;
  repeated Vector3 prediction = 8;
}

message Object {
  int32 id = 1;
  BoundingBox bbox = 2;
  label_types.LabelType label = 3;
  TrackingInfo track = 101;
  bytes points = 102;                    // Data type of points is Vector3f(12bytes). When you parsing this data, please be aware of this.
}

message Zone {
  int32 id = 1;
  PolygonBox polygon_box = 2;
  repeated int32 object_ids = 3;
  bool is_exclusion = 4;
}

// Actual message using in SENSR SDK
message OutputMessage {
  google.protobuf.Timestamp time_stamp = 1;
  repeated Object objects = 2;
  repeated Zone region_of_interest = 101;
  bytes ground_points = 102;              // Data type of points is Vector3f(12bytes). When you parsing this data, please be aware of this.
}

// Deprecated Message type
message point_list {
  int32 size = 1 [deprecated=true];
  bytes points = 2 [deprecated=true];
}

message point_cloud {
  point_list ground_points = 1 [deprecated=true];
  point_list environment_points = 2 [deprecated=true];
}

message time_stamp {
  int64 seconds = 1 [deprecated=true];
  int32 nano_seconds = 2 [deprecated=true];
}